### PR TITLE
Add torch._foreach_maximum(TensorList, TensorList) & torch._foreach_minimum(TensorList, TensorList) APIs

### PR DESCRIPTION
**Pull request by izdeby**
- Adding torch._foreach_maximum(TensorList, TensorList) API
- Adding torch._foreach_minimum(TensorList, TensorList) API
- Updated Adam/AdamW optimizers 

Tested via unit tests

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachPointwiseOp.cu [hunk 1]
```diff
@@ -74,7 +76,40 @@ void foreach_tensor_##NAME##_cuda_(TensorList input, TensorList tensors1, Tensor
...
+    tensor_lists.emplace_back(tensors1.vec());
+    tensor_lists.emplace_back(tensors2.vec());
+    tensor_lists.emplace_back(std::move(vec_res));
+    auto res = std::max<int> (2, 3); // WORKS
+    auto res2 = std::max<float> (2, 3); // WORKS
+    
+    
+    AT_DISPATCH_ALL_TYPES_AND(kHalf, tensors1[0].scalar_type(), "foreach_pointwise_op_cuda", [&]() {
+        using opmath_t = get_opmath_t<scalar_t>::opmath_t;
+        auto op = []  __device__ (opmath_t a, opmath_t b) -> opmath_t { return std::max(a, b); };
```
Comment (mcarilli) [inline aten/src/ATen/native/cuda/ForeachPointwiseOp.cu:106]: Maybe replace `__device__` with GPU_LAMBDA?

Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachPointwiseOp.cu:106]: Done. the error is still there -` terminated by signal SIGSEGV (Address boundary error)`

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 2]
```diff
@@ -656,6 +656,74 @@ struct PointwiseOpFunctor {
         }
 };
 
+template<typename T>
+struct BinaryOpListFunctor {
```
Comment (izdeby) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:660]: I promise i will refactor the whole file one day.

Comment (ngimel) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:660]: Ok, for now maybe one line comments for each variant would be helpful (unless you plan to refactor soon). 
E.g. BinaryOpListFunctor - 2 inputs, one out of place output, PointwiseOpFunctor - 3 inputs, one in-place output etc.

File: aten/src/ATen/native/cuda/ForeachPointwiseOp.cu [hunk 3]
```diff
@@ -74,7 +73,37 @@ void foreach_tensor_##NAME##_cuda_(TensorList input, TensorList tensors1, Tensor
...
+        vec_res.emplace_back(at::native::empty_like(t));
+    }
+
+    tensor_lists.emplace_back(tensors1.vec());
+    tensor_lists.emplace_back(tensors2.vec());
+    tensor_lists.emplace_back(std::move(vec_res));
+    
+    AT_DISPATCH_ALL_TYPES_AND(kHalf, tensors1[0].scalar_type(), "foreach_pointwise_op_cuda", [&]() {
+        using opmath_t = get_opmath_t<scalar_t>::opmath_t;
+        auto op = []  GPU_LAMBDA (opmath_t a, opmath_t b) -> opmath_t { return std::max(a, b); };
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachPointwiseOp.cu:100]: I'm afraid std::max won't properly propagate nans, and you want behavior to agree with eager mode. Please see a functor that eager max uses for it. Would also be good to add tests with nans.

File: test/test_optim.py [hunk 4]
```diff
@@ -334,7 +334,7 @@ def test_multi_tensor_optimizers(self):
             ((optim.Adadelta, optim._multi_tensor.Adadelta), dict(weight_decay=1)),
         ]
 
-        kIterations = 1001
+        kIterations = 11
```
Comment (ngimel) [inline test/test_optim.py:337]: good!

**Review by ngimel — COMMENTED** [review]
Would also make sense to add `min`, even though it is not used in existing optimizers.

File: test/test_optim.py (hunk 4 above)
Comment (mcarilli) [reply test/test_optim.py:337]: whew, there's the 10 min tests.

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 2 above)
Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:660]: ill refactor it soon

File: aten/src/ATen/native/cuda/ForeachPointwiseOp.cu (hunk 3 above)
Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachPointwiseOp.cu]: Added two [tests](https://github.com/pytorch/pytorch/pull/45692/files#diff-b300fed36e386acda9146c82f11d6090R199) and they are passing. am i missing something?

Comment (ngimel) [reply aten/src/ATen/native/cuda/ForeachPointwiseOp.cu]: Yeah, using std::min/std::max is unsafe, see e.g. here https://stackoverflow.com/questions/39919069/why-does-release-debug-have-a-different-result-for-stdmin. Tests might be passing in some circumstances and failing in others.

**Review by ngimel — COMMENTED** [review]
Another minor change - pointwise min/max are being deprecated in favor of minimum/maximum, so can you please call this function _foreach_maximum to disambiguate with min/max reduction?

File: aten/src/ATen/native/cuda/ForeachPointwiseOp.cu [hunk 5]
```diff
@@ -74,7 +75,46 @@ void foreach_tensor_##NAME##_cuda_(TensorList input, TensorList tensors1, Tensor
...
+                                                                                                           \
+    tensor_lists.emplace_back(tensors1.vec());                                                             \
+    tensor_lists.emplace_back(tensors2.vec());                                                             \
+    tensor_lists.emplace_back(std::move(vec_res));                                                         \
+                                                                                                           \
+    AT_DISPATCH_ALL_TYPES_AND(kHalf, tensors1[0].scalar_type(), "foreach_maximum_minimum_op_cuda", [&]() { \
+        using opmath_t = get_opmath_t<scalar_t>::opmath_t;                                                 \
+        auto op = []  GPU_LAMBDA (opmath_t a, opmath_t b) -> opmath_t {                                    \
+            opmath_t c = a OP b ? a : b;                                                                   \
+            if (_isnan(a)) {                                                                               \
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachPointwiseOp.cu:105]: nit:
```
return _isnan(a) ? a : c;
```

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**